### ISSUE TITLE
wrapper for calls to ui._raw_main in test_ui.py

### ISSUE THREAD

**Issue by jrobeson** [beetbox/beets#2138]
Currently we're unable to use `run_command` from `helper.py` due to the usage of `self.lib` as seen here https://github.com/beetbox/beets/blob/master/test/helper.py#L424 . Using `run_command` directly ends up causing a problem due to the autodetection of the config filename here: https://github.com/beetbox/beets/blob/master/beets/ui/__init__.py#L1161 Should we create another version of `run_command` for `test_ui.py` or change how `run_command` works? or something else?

This would really help make the py3 changes a lot simpler (due to bytes vs string arguments), and seems like a good idea in any case.

**Comment by sampsyo** [comment]
Please forgive me for being slow on the uptake, but I don't quite see the root cause here. Is the problem that `argv` is mixing bytes and stir? Or that, on Python 3, `args` can't be bytes at all?

It might be the jetlag talking here, but I think I might be on the *brink* of understanding...

**Comment by jrobeson** [comment]
args can't be bytes when using the the default stdio stuff.

**Comment by jrobeson** [comment]
I should definitely mention that subprocess in/out is still bytes, while the args must be strings.

**Comment by jrobeson** [comment]
I was trying to avoid a conversation about where we're going with this in this particular ticket, as it deserves a whole post of it's own. I just thought that providing more reusable `run_command` was a goal useful it's own right while still allowing me to clean up my py3 PR.

**Comment by sampsyo** [comment]
> args can't be bytes when using the the default stdio stuff.

OK, I might need one more step to understand… what do args have to do with stdin and stdout?

Or, perhaps a more relevant question: what's the exact point in the current workflow that requires that arguments be Unicode strings? Is it some interaction with a standard library module? If so, which module, and where do we call it? (Maybe we can bypass that call to sidestep the requirement.)

If I'm understanding correctly, maybe the right thing to do is to rethink the way tests invoke commands. Most commands use `decargs` to retrieve the arguments as Unicode strings, but sometimes I imagine they want the raw bytes (e.g., when a filename is an argument). So with that in mind, we need to decide what tests should be able to supply—should they be able to send Unicode arguments, which we'll then encode so they get decoded correctly? Should they be able to provide bytes (seems necessary for the filename case)?

**Comment by jrobeson** [comment]
if argv itself is contained of strings, where are the bytes coming from?

```python
#!/usr/bin/python3

import sys

for a in sys.argv:
    print(type(a))

```



```sh

./test_args.py foo bar baz 1
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
```

**Comment by jrobeson** [comment]
Of course, we can still read bytes from stdin via `sys.stdin.buffer`, and write them to stdout via `sys.stdout.buffer if that wasn't clear.`

**Comment by jrobeson** [comment]
You can also see that optparse never deals with anything but strings either: https://hg.python.org/cpython/file/3.4/Lib/optparse.py#l1406

**Comment by sampsyo** [comment]
Right, I understand that `sys.argv` consists of Unicode strings on Python 3—I was just hoping we could bypass that altogether for our handling. (Once again, Python 3's policy is making things seem simpler but actually making it harder for us to write correct software…)

Thanks for clarifying that we're out of luck with optparse; I suppose that means we need to convert everything to strings. So maybe the right solution is to do `b.decode(_arg_encoding(), 'surrogateescape')` to convert byte arguments from the tests to strings?

**Comment by jrobeson** [comment]
that gets back to the original qustion though. a wrapper for for the calls to `_raw_main`. I'd prefer not to add that decoding to before every call.  I'd prefer to use `run_command` but that's not possible with the way due to the way it checks `self.lib`

**Comment by jrobeson** [comment]
fixed in 9ab9166c8d5859fad53fd8a3ff422c6ff317b8b3
